### PR TITLE
[WIP] NF: Wigner-D Rotation Functions

### PR DESCRIPTION
**Pull request by aayush9400**
The `wigner_rotation` performs a Wigner rotation on a given input array 'x', representing spherical harmonics coefficients. This rotation is defined by Euler angles alpha, beta, and gamma.

**Method Signature:**

```def wigner_rotation(signal: ndarray, alpha: int, beta: int, gamma: int) -> numpy.ndarray of same shape as image```

**Example Usage**
```
import numpy as np
from keras.datasets import mnist
from utils_wigner import create_spherical, plot_spherical_img
from dipy.reconst.shm import wigner_rotation

def rotation(x, a, b, c):
    shape = x.shape
    multiples = np.append(np.ones_like(shape), shape[-1])
    x = np.tile(np.expand_dims(x, axis=-1), multiples)
    x = wigner_rotation(x, a, b, c)  
    return x[..., 0]


if __name__ == '__main__':
    (train_X, train_y), (test_X, test_y) = mnist.load_data()
    index = 0
    arr = train_X[index]    
    print("projecting {0} data image".format(train_y[index]))
    og_img = create_spherical(arr)[0]
    
    rot_img = rotation(og_img.astype(np.float32), 0, 0, 10)
    rot_img[rot_img<0] = 0
    plot_spherical_img(og_img, rot_img)
```

### CODE REVIEW

File: dipy/reconst/shm.py [hunk 1]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
     return out_sh_coeffs
+
+
+def _setup_wigner_rotation(b, alpha, beta, gamma):
+    '''
+    Returns an array of all possible wigner-D matrices for a range of quantum numbers (l)
+
+    Parameters
+    ----------
+    b : int
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1587]: we avoid single character as a parameter. can you rename it ?

File: dipy/reconst/shm.py [hunk 2]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
+        The upper limit for the range of quantum numbers. The function generates Wigner-D
+        matrices for each quantum number from 0 to b-1.
+    alpha : float
+        The first Euler angle, representing a rotation around the z-axis.
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1591]: `z-axis` ?

File: dipy/reconst/shm.py [hunk 3]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
         raise ValueError("Invalid basis name.")
 
     return out_sh_coeffs
+
+
+def _setup_wigner_rotation(b, alpha, beta, gamma):
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1581]: alpha, beta, gamma could be 1 argument. Why did you choose to split it ?

File: dipy/reconst/shm.py [hunk 4]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
+    alpha : float
+        The first Euler angle, representing a rotation around the z-axis.
+    beta : float
+        The second Euler angle, representing a rotation around the y-axis.
+    gamma : float
+        The third Euler angle, representing a rotation around the z-axis.
+
+    Returns
+    -------
+    Us : list of ndarray
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1599]: we avoid list of ndarray.  Can you just increase the dimension of your final array?

File: dipy/reconst/shm.py [hunk 5]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
+    - The 'Condon-Shortley' phase convention is used.
+    '''
+    Us = [wigner_D_matrix(l, alpha, beta, gamma,
+                          field='complex', normalization='quantum', order='centered', condon_shortley='cs')
+          for l in range(b)]
+    Us = [Us[l].astype(np.complex64).view(np.float32).reshape((2 * l + 1, 2 * l + 1, 2)) for l in range(b)]
+    return Us
+
+
+def wigner_rotation(x, alpha, beta, gamma):
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1614]: same comment as above

File: dipy/reconst/shm.py [hunk 6]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
+    '''Performs a Wigner rotation on a given input array 'x', representing spherical
+    harmonics coefficients. This rotation is defined by Euler angles alpha, beta,
+    and gamma. The function first sets up the Wigner-D matrices for the specified
+    range of quantum numbers, then applies a rotation in the spherical harmonic domain.
+
+    Parameters
+    ----------
+    x : ndarray
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1622]: same comment as above

File: dipy/reconst/shm.py [hunk 7]
```diff
@@ -1575,3 +1576,95 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
+    z : ndarray
+        The rotated array, with the same shape as the input 'x', representing the
+        spherical harmonics coefficients after the Wigner rotation
+    
+    References
+    ----------
+    .. [1] Taco S. Cohen, Mario Geiger, Jonas Köhler, Max Welling, Spherical CNNs. 
+           International Conference on Learning Representations (ICLR), 2018.
+
+    '''
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1644]: same comment as above

File: dipy/reconst/wigner.py [hunk 8]
```diff
@@ -0,0 +1,618 @@
+import os
+import numpy as np
+import collections
+from scipy.linalg import block_diag
+from functools import lru_cache
+from dipy.data import get_fnames
+
+fetch_Jmat_path = get_fnames('wigner_jmat')
+print('fetched ' + fetch_Jmat_path)
+Jd = np.load(fetch_Jmat_path, allow_pickle=True)
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:10]: - can you remove the print ?
- it should not be global. please, create a decorator or just a function

File: dipy/reconst/wigner.py [hunk 9]
```diff
@@ -0,0 +1,618 @@
...
+    if b_out is None:
+        b_out = b_in
+    batch_size = x.shape[:-3]
+
+    # [batch, beta, alpha, gamma]
+    x = np.reshape(x, (-1, 2 * b_in, 2 * b_in, 2 * b_in))
+    '''
+    :param x: [batch, beta, alpha, gamma] (nbatch, 2 b_in, 2 b_in, 2 b_in)
+    :return: [l * m * n, batch, complex] (b_out (4 b_out**2 - 1) // 3, nbatch, 2)
+    '''
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:470]: Can you remove this ?

File: dipy/reconst/wigner.py [hunk 10]
```diff
@@ -0,0 +1,618 @@
...
+    -------
+    ndarray
+        The spectral representation of the input signal with dimensions 
+        [l * m * n, ..., complex], where 'l * m * n' represents the spectral coefficients 
+        and '...' are the batch dimensions. The output is complex-valued.
+    '''
+    b_in = x.shape[-1] // 2
+    assert x.shape[-1] == 2 * b_in
+    assert x.shape[-2] == 2 * b_in
+    assert x.shape[-3] == 2 * b_in
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:460]: Can you improve this test for better messaging for the user?

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @aayush9400,

Thank you for doing this! See below an initial review. I still need to go deeper in this review

File: dipy/reconst/shm.py (hunk 4 above)
Comment (aayush9400) [reply dipy/reconst/shm.py:1599]: For this array indexing is important, so a list of ndarray makes sense. Every index has a specific ndarray associated with it. Other alternatives could be to use hash maps.

File: dipy/reconst/shm.py (hunk 3 above)
Comment (aayush9400) [reply dipy/reconst/shm.py:1581]: They can be converted into a list but these variables are user inputs so for better readability and understanding they are separate. Should I convert them to a list?

File: dipy/data/fetcher.py [hunk 11]
```diff
@@ -691,6 +692,14 @@ def fetcher():
      'c1499005baccfab865ce38368d7a4c7f'],
     doc="Download Bundle Warp dataset")
 
+fetch_wigner_Jmat = _make_fetcher(
+    "fetch_wigner_Jmat",
+    pjoin(dipy_home, 'wigner'),
+    'https://ndownloader.figshare.com/files/',
+    ['43638069'],
+    ['J_dense_0-150.npy'],
```
[the remark this answers is not in the document]
Comment (aayush9400) [reply dipy/data/fetcher.py:700]: Updated and fixed!

File: dipy/reconst/wigner.py (hunk 8 above)
Comment (aayush9400) [reply dipy/reconst/wigner.py:10]: created a new function

File: dipy/data/fetcher.py [hunk 12]
```diff
@@ -691,6 +691,15 @@ def fetcher():
      'c1499005baccfab865ce38368d7a4c7f'],
     doc="Download Bundle Warp dataset")
 
+fetch_wigner_Jmat = _make_fetcher(
```
Comment (skoudoro) [inline dipy/data/fetcher.py:694]: `jmat` lowercase to avoid any confusion and respect the standard of the other fetcher

File: dipy/data/fetcher.py [hunk 13]
```diff
@@ -691,6 +691,15 @@ def fetcher():
...
+    "fetch_wigner_Jmat",
```
Comment (skoudoro) [inline dipy/data/fetcher.py:695]: same as above

File: dipy/reconst/shm.py (hunk 4 above)
Comment (skoudoro) [reply dipy/reconst/shm.py:1599]: I understand. That's why I propose you to increase the dimension. 

from `[..., beta, alpha, gamma]`, create new array with `[..., beta, alpha, gamma, index]` where instead of list of ndarray which has memory issues.

File: dipy/reconst/shm.py [hunk 14]
```diff
@@ -1575,3 +1576,100 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
         raise ValueError("Invalid basis name.")
 
     return out_sh_coeffs
+
+
+def _setup_wigner_rotation(quant_order, alpha, beta, gamma):
+    """Returns an array of all possible wigner-D matrices for a range
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1582]: - "Return" instead of "Returns" ( imperative mood)
- It should be one-line summary. we follow Numpy docstring standard: https://numpydoc.readthedocs.io/en/latest/format.html

File: dipy/reconst/shm.py [hunk 15]
```diff
@@ -1575,3 +1576,100 @@ def convert_sh_to_legacy(sh_coeffs, sh_basis, full_basis=False):
...
+          for l in range(quant_order)]
+    return Us
+
+
+def wigner_rotation(signal, alpha, beta, gamma):
+    """Performs a Wigner rotation on a given input array 'x', representing 
+    spherical harmonics coefficients. This rotation is defined by Euler angles 
+    alpha, beta, and gamma. The function first sets up the Wigner-D matrices for
+    the specified range of quantum numbers, then applies a rotation in the 
+    spherical harmonic domain.
```
Comment (skoudoro) [inline dipy/reconst/shm.py:1624]: - "Return" instead of "Returns" ( imperative mood)
- It should be one-line summary. we follow Numpy docstring standard: https://numpydoc.readthedocs.io/en/latest/format.html

File: dipy/reconst/wigner.py [hunk 16]
```diff
@@ -0,0 +1,660 @@
...
+    fetch_Jmat_path = get_fnames('wigner_jmat')
+    Jd = np.load(fetch_Jmat_path, allow_pickle=True)
+    return Jd
+
+
+def z_rot_mat(angle, l):
+    """
+    Create the matrix representation of a z-axis rotation by the given angle,
+    in the irrep l of dimension 2 * l + 1, in the basis of real centered
+    spherical harmonics.
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:19]: same comment as above

File: dipy/reconst/wigner.py [hunk 17]
```diff
@@ -0,0 +1,660 @@
...
+
+    output = np.reshape(output, [*batch_size, 2 * b_out, 2 * b_out, 2 * b_out])
+    return output
+
+
+def complex_mm(x, y, conj_x=False, conj_y=False):
+    """Perform matrix multiplication (mm) of complex matrices. Given two complex
+      matrices 'x' and 'y', this function computes their product. The function 
+      also supports optional conjugation of either or both input matrices before
+      multiplication.
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:620]: same comment as above

File: dipy/reconst/wigner.py [hunk 18]
```diff
@@ -0,0 +1,660 @@
...
+
+    ifft_output = np.fft.ifft2((output[..., 0] + 1j * output[..., 1]), 
+                               axes=[2, 3]) * float(output.shape[-2]) ** 2
+    output = np.real(ifft_output)
+
+    output = np.reshape(output, [*batch_size, 2 * b_out, 2 * b_out, 2 * b_out])
+    return output
+
+
+def complex_mm(x, y, conj_x=False, conj_y=False):
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:616]: I do not think this function should be in this module.

File: dipy/reconst/wigner.py [hunk 19]
```diff
@@ -0,0 +1,660 @@
...
+    # [l * m * n, ..., complex]
+    output = np.reshape(output, (-1, *batch_size, 2))
+    return output
+
+
+def so3_rifft(x, for_grad=False, b_out=None):
+    """
+    Perform an inverse real Fourier transform on the SO(3) group. This function 
+    transforms a spectral representation back into its signal representation on 
+    the 3D rotation group using the inverse Wigner-D functions.
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:550]: same as above

File: dipy/reconst/wigner.py [hunk 20]
```diff
@@ -0,0 +1,660 @@
...
+        bandwidth.
+
+    Returns
+    -------
+    ndarray
+        The signal representation of the spectral input with dimensions 
+        [..., beta, alpha, gamma], where 'beta', 'alpha', and 'gamma' are the 
+        Euler angles defining the rotation and '...' are the batch dimensions.
+    """
+    assert x.shape[-1] == 2
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:573]: Can you raise an exception instead using a good message?

File: dipy/reconst/wigner.py [hunk 21]
```diff
@@ -0,0 +1,660 @@
...
+    nspec = x.shape[0]
+    b_in = round((3 / 4 * nspec) ** (1 / 3))
+    assert nspec == b_in * (4 * b_in**2 - 1) // 3
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:576]: Can you raise an exception instead using a good message?

File: dipy/reconst/wigner.py [hunk 22]
```diff
@@ -0,0 +1,660 @@
...
+        dss.append(ds)
+    dss = np.stack(dss)  # [beta, l * m * n]
+    return dss
+
+
+def so3_rfft(x, for_grad=False, b_out=None):
+    """
+    Perform a real Fourier transform on the SO(3) group. This function 
+    transforms a signal defined on the 3D rotation group to its spectral 
+    representation using the Wigner-D functions.
```
Comment (skoudoro) [inline dipy/reconst/wigner.py:464]: same comment as above

**Review by skoudoro — COMMENTED** [review]
Hi @aayush9400,

- Can you address all the pep8 issues?
- Can you fix the merge conflict on your branch. We can not test it for now.
- Have you look at [`dipy.core.geometry`](https://github.com/dipy/dipy/blob/master/dipy/core/geometry.py#L548) ?  I feel like some function are duplicated/reimplemented. I need look deeper into that. I will do it after you fix the conflict.  I will also look at `nibabel.eulerangles`. if it is ok, I would recommend to move some function in `dipy.core.geometry`.
- Every function should be tested individually. So you need a `test_wigner.py` which test every single function. it makes everything more robust and help for the review.
- See below for additional comments

Thanks for looking into that and let me know when I can do a deeper review and test it.

**Comment by Garyfallidis** [comment]
Okay. Good progress. Next step is adding simple to follow examples.

**Comment by skoudoro** [comment]
Hi @aayush9400,

What is the status of this work/PR. Can you rebase it and address all the conflict?

Thank you in advance for your feedback

**Comment by skoudoro** [comment]
Hi @aayush9400,

Same questions as above.

**Comment by skoudoro** [comment]
I got some feedback. For now, Closing the PR.

Feel free to reopen it when it is ready for review.

Thanks @aayush9400 !!
